src/Database/LSMTree/Internal/Run/Construction.hs

```diff
@@ -41,20 +46,166 @@ module Database.LSMTree.Internal.Run.Construction (
+unsafeFinalise ::
+     MRun s
+  -> ST s ( Maybe (PageAcc, [Index.Chunk])
+          , Maybe Index.Chunk
+          , Index.FinalChunk
+          , Bloom SerialisedKey
+          , CompactIndex
+          )
+unsafeFinalise mrun@MRun {..} = do
+    mpage <- yield mrun
+    (mchunk, fchunk) <- Index.unsafeEnd mindex
+    storeChunk mrun mchunk
+    bloom <- Bloom.freeze mbloom
+    allChunks <- getAllChunks mrun
+    let index = Index.fromChunks allChunks fchunk
+    pure (mpage, mchunk, fchunk, bloom, index)
+
+-- | Add a serialised k\/op pair with an optional blob reference. Use only for
+-- entries that are fully in-memory. Otherwise, use 'addMultiPageInChunks'.
+addFullKOp ::
+     MRun s
+  -> SerialisedKey
+  -> Entry RawValue BlobRef
+  -> ST s (Maybe (PageAcc, [Index.Chunk]))
+addFullKOp  mrun@MRun{..}k e = do
+    Bloom.insert mbloom k
+    p <- readSTRef currentPageRef
+    case paAddElem rangeFinderPrecision k e p of
+      Nothing -> do
+        (pb, cs) <- unsafeYieldAlways mrun
+        writeSTRef currentPageRef $! paSingleton k e
+        pure $ Just (pb, cs)
+      Just p' -> do
+        writeSTRef currentPageRef $! p'
+        pure Nothing
+
+-- | __For multi-page values only during run merging:__ add the first chunk of a
+-- serialised k\/op pair with an optional blob reference. The caller of this
+-- function is responsible for copying the remaining chunks to the output file.
+--
+-- This will throw an impure exception if the result of adding the k\/op pair is
+-- not exactly the size of a disk page.
+--
+-- During a merge, raw values that span multiple pages may not be read into
+-- memory as a whole, but in chunks. If you have access to the full bytes of a
+-- raw value, use 'addFullKOp' instead.
+addChunkedKOp ::
+     MRun s
+  -> SerialisedKey
+  -> Entry RawValue BlobRef
+  -> Word32
+  -> ST s (Either PageAcc (PageAcc, PageAcc), [Index.Chunk])
+addChunkedKOp mrun@MRun{..} k e nOverflow
+  | pageSizeNumBytes (psSingleton k e) == 4096 = do
+      mp <- yield mrun
+
+      Bloom.insert mbloom k
+      cs' <- Index.append (Index.AppendMultiPage k nOverflow) mindex
+      storeChunks mrun cs'
+      let p' = paSingleton k e
+
+      case mp of
+        Nothing      -> pure (Left p', cs')
+        Just (p, cs) -> pure (Right (p, p'), cs' <> cs)
+  | otherwise = error "addMultiPage: expected a page of 4096 bytes"
+
+-- | Yield the current page if it is non-empty. New chunks are recorded in the
+-- mutable run, and the current page is set to empty.
+yield :: MRun s -> ST s (Maybe (PageAcc, [Index.Chunk]))
+yield mrun@MRun{..} = do
+    p <- readSTRef currentPageRef
+    if paIsEmpty p then
+      pure Nothing
+    else do
+      cs <- Index.append (unsafeMkAppend p) mindex
+      storeChunks mrun cs
+      writeSTRef currentPageRef $! paEmpty
+      pure $ Just (p, cs)
+
+-- | Like 'yield', but can fail if the page is empty.
+unsafeYieldAlways :: MRun s -> ST s (PageAcc, [Index.Chunk])
```

Comment:
As I noted before, are we sure it's worth adding an almost identical, but unsafe version just to avoid a single `if`? It also avoids returning a `Maybe`, but the only caller immediately wraps the result in `Just` anyways.

Reply:
I personally don't mind a little bit of code duplication, but I can remove it

Reply:
But currently instead of allocating the Maybe here, `addFullKOp` just wraps the result in a Maybe, which then could go away. :shrug: 

I don't think it really matters for performance either way, so I'd prefer to keep the code simpler.

Reply:
Oops, you're right